File: frontend/src/lib/updates/socket.ts
// The one socket, and the loop that keeps it up.
//
// It reuses the same line chooser the live terminal uses (connectOverLines), so this connection is
// not pinned to whichever route served the page — and, like that one, it authenticates with a token
// in the query string rather than a cookie, because a WebSocket handshake is not subject to CORS.
//
// A heartbeat is here for one specific failure and not for any other: a half-open socket, where the
// TCP connection is gone but no close event ever fires. That is the only thing a heartbeat can
// detect. It cannot detect the failure this system is actually prone to — an event that was never
// published at all — and nothing here should be read as covering that. The polling left in place
// upstream is what covers that, deliberately.

import { connectOverLines } from "@micro-teams/multipath";
import { lineManager, MT_PATH } from "@/lib/lines";
import { getNtAccessToken } from "@/lib/mtApi";
import { parseFrame } from "@/lib/updates/protocol";
import { UpdatesStore } from "@/lib/updates/store";

/** How long the socket may stay silent before we assume it is a corpse and reconnect. */
const SILENCE_MS = 45_000;
/** How often we prod it. Well inside SILENCE_MS so a healthy socket never trips the check. */
const PING_MS = 20_000;

export interface UpdatesSocket {
  close: () => void;
}

/**
 * Open the updates socket and pump it into `store`. Returns a handle whose `close` stops everything
 * including the reconnect loop.
 */
export function openUpdatesSocket(store: UpdatesStore): UpdatesSocket {
  let closed = false;
  let ws: WebSocket | null = null;
  let lastHeard = Date.now();

  const ping = window.setInterval(() => {
    if (closed) return;
    const socket = ws;
    if (!socket || socket.readyState !== WebSocket.OPEN) return;
    if (Date.now() - lastHeard > SILENCE_MS) {
      // Nothing has arrived for a long time, not even our own pong. Close it and let the reconnect
      // loop dial again — a socket that is dead but not closed is the one thing this timer exists
      // to notice.
      try {
        socket.close();
      } catch {
        /* already gone */
      }
      return;
    }
    try {
      socket.send(JSON.stringify({ t: "ping" }));
    } catch {
      /* the close handler will deal with it */
    }
  }, PING_MS);

  // Coming back to the tab refetches everything once — see UpdatesStore.refocused for why this
  // stays forever even though the polls are gone.
  const onVisible = () => {
    if (!closed && document.visibilityState === "visible") store.refocused();
  };
  document.addEventListener("visibilitychange", onVisible);
  window.addEventListener("focus", onVisible);

  const connection = connectOverLines({
    lines: () => lineManager.ranked(),
    path: `${MT_PATH}/updates${tokenQuery()}`,
    createSocket: (url) => {
      const socket = new WebSocket(url);
      ws = socket;
      socket.onopen = () => {
        lastHeard = Date.now();
        store.connected({
          send: (frame) => {
            try {
              socket.send(JSON.stringify(frame));
            } catch {
              /* dropped while closing; the next connect resubscribes everything */
            }
          },
        });
      };
      socket.onmessage = (ev) => {
        lastHeard = Date.now();
        if (typeof ev.data !== "string") return;
        const frame = parseFrame(ev.data);
        if (frame) store.handle(frame);
      };
      return socket;
    },
    onClose: () => {
      store.disconnected();
    },
  });

  return {
    close: () => {
      closed = true;
      document.removeEventListener("visibilitychange", onVisible);
      window.removeEventListener("focus", onVisible);
      window.clearInterval(ping);
      store.disconnected();
      connection.close();
      try {
        ws?.close();
      } catch {
        /* already gone */
      }
    },
  };
}

function tokenQuery(): string {
  const token = getNtAccessToken();
  return token ? `?token=${encodeURIComponent(token)}` : "";
}

File: app/lib/src/common/updates/socket.dart
/// The one socket, and the loop that keeps it up.
///
/// A heartbeat is here for one specific failure and not for any other: a half-open socket, where
/// the connection is gone but no close event ever fires. That is the only thing a heartbeat can
/// detect. It cannot detect the failure this system is actually prone to — an event that was never
/// published at all — and nothing here should be read as covering that. The periodic state frame
/// from the server, compared against each subscriber's digest, is what covers that.
///
/// On a phone there is a second reason this file is careful: the OS suspends the process without
/// telling anyone, and the socket comes back as a corpse. Every path out of that ends in
/// [UpdatesStore.refocused] or a reconnect, both of which cost one refetch and cannot show
/// stale data.
library;

import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

import 'protocol.dart';
import 'store.dart';

/// How long the socket may stay silent before we assume it is a corpse and reconnect.
const Duration _silence = Duration(seconds: 45);

/// How often we prod it. Well inside [_silence] so a healthy socket never trips the check.
const Duration _pingEvery = Duration(seconds: 20);

/// Reconnect backoff. Capped low: this socket carries nothing but "go and look", so dialling
/// again cheaply is better than being clever.
const Duration _minRetry = Duration(seconds: 1);
const Duration _maxRetry = Duration(seconds: 20);

class UpdatesSocket {
  UpdatesSocket({required UpdatesStore store, required this.url})
    : _store = store;

  final UpdatesStore _store;

  /// Called for each dial, so the token is always the current one — a socket that reconnects with
  /// yesterday's token is refused, and looks exactly like a server that has gone quiet.
  final String Function() url;

  /// Told when a dial succeeds and when it ends, so the line policy can learn which lines can hold
  /// a stream at all — a different question from which answers requests quickly. Optional because
  /// the terminal's socket and the tests do not have a policy to inform.
  void Function()? onOpened;
  void Function()? onClosed;

  WebSocketChannel? _channel;
  StreamSubscription<Object?>? _messages;
  Timer? _ping;
  Timer? _retry;
  DateTime _lastHeard = DateTime.now();
  Duration _backoff = _minRetry;
  bool _closed = false;

  void start() {
    _closed = false;
    _dial();
    _ping = Timer.periodic(_pingEvery, (_) => _heartbeat());
  }

  /// The app came back to the foreground.
  void resumed() {
    if (_closed) return;
    _store.refocused();
    // A socket that was suspended may be dead without knowing it. Prod it now rather than waiting
    // out the silence window, because the human is looking at the screen right now.
    _heartbeat();
  }

  void close() {
    _closed = true;
    _ping?.cancel();
    _retry?.cancel();
    unawaited(_messages?.cancel());
    _store.disconnected();
    _channel?.sink.close();
    _channel = null;
  }

  void _dial() {
    if (_closed) return;
    try {
      final channel = WebSocketChannel.connect(Uri.parse(url()));
      onOpened?.call();
      _channel = channel;
      _lastHeard = DateTime.now();
      _messages = channel.stream.listen(
        _onMessage,
        onDone: () {
          onClosed?.call();
          _onClosed();
        },
        onError: (Object _) => _onClosed(),
        cancelOnError: true,
      );
      _store.connected(_ChannelTransport(channel));
      _backoff = _minRetry;
    } catch (_) {
      _onClosed();
    }
  }

  void _onMessage(Object? data) {
    _lastHeard = DateTime.now();
    if (data is! String) return;
    final frame = parseFrame(data);
    if (frame != null) _store.handle(frame);
  }

  void _onClosed() {
    if (_closed) return;
    _store.disconnected();
    _channel = null;
    _retry?.cancel();
    _retry = Timer(_backoff, _dial);
    final next = _backoff * 2;
    _backoff = next > _maxRetry ? _maxRetry : next;
  }

  void _heartbeat() {
    if (_closed) return;
    final channel = _channel;
    if (channel == null) return;
    if (DateTime.now().difference(_lastHeard) > _silence) {
      // Nothing has arrived for a long time, not even our own pong. Close it and let the reconnect
      // loop dial again — a socket that is dead but not closed is the one thing this timer exists
      // to notice.
      channel.sink.close();
      _onClosed();
      return;
    }
    try {
      channel.sink.add(jsonEncode(const PingFrame().toJson()));
    } catch (_) {
      _onClosed();
    }
  }
}

class _ChannelTransport implements UpdatesTransport {
  _ChannelTransport(this._channel);

  final WebSocketChannel _channel;

  @override
  void send(ClientFrame frame) {
    try {
      _channel.sink.add(jsonEncode(frame.toJson()));
    } catch (_) {
      // Dropped while closing; the next connect resubscribes everything.
    }
  }
}
